File: C++/Rectangle.cpp
#include "Rectangle.h"

Rectangle::Rectangle(std::string name, std::initializer_list<Vertex> ilist) : Shape(name, ilist) {
	if (ilist.size() != 4)
		throw std::invalid_argument("Square requires 4 vertices");
	checkAndSet();
}

Rectangle::Rectangle(std::string name, Vertex lb_vtx, _Vertex_t sW, _Vertex_t sH) :
	Shape(name, { lb_vtx,
		  {lb_vtx.get_x(), lb_vtx.get_y() + sH},
		  {lb_vtx.get_x() + sW, lb_vtx.get_y() + sH},
		  {lb_vtx.get_x() + sW, lb_vtx.get_y()}
		  }),
	side_w(sW),
	side_h(sH)
{
	if (sW <= 0 or sH <= 0)
		throw std::invalid_argument("Side lengths must be > 0");
}

void Rectangle::checkAndSet() {
	if (((*this)[2] - (*this)[0]).abs() != ((*this)[3] - (*this)[1]).abs())
		throw std::invalid_argument("{}: The given points don't form a rectangle");

	side_h = ((*this)[1] - (*this)[0]).hypot();
	side_w = ((*this)[2] - (*this)[1]).hypot();
}

_Vertex_t Rectangle::get_side_w() const {
	return side_w;
}

_Vertex_t Rectangle::get_side_h() const {
	return side_h;
}

Rectangle::Rectangle(Rectangle& other) :
	Shape(other.name, other.vertices),
	side_h(other.side_h),
	side_w(other.side_w) {}

Rectangle::Rectangle(Rectangle&& other) noexcept :
	Shape(std::move(other.name), std::move(other.vertices)),
	side_h(std::exchange(other.side_h, 0)),
	side_w(std::exchange(other.side_w, 0)) {}

Rectangle& Rectangle::operator=(Rectangle& other) {
	name = other.name;
	vertices = other.vertices;
	side_h = other.side_h;
	side_w = other.side_w;

	return *this;
}

Rectangle& Rectangle::operator=(Rectangle&& other) noexcept {
	name = std::move(other.name);
	vertices = std::move(other.vertices);
	side_h = std::exchange(other.side_h, 0);
	side_w = std::exchange(other.side_w, 0);
	return *this;
}

double Rectangle::square() {
	return side_h * side_w;
}

nlohmann::json Rectangle::makeJson() {
	nlohmann::json json;
	json["type"] = "Rectangle";
	json["name"] = name;
	for (auto& vtx : *this)
		json["vertices"].push_back({ { "x", vtx.get_x() }, { "y", vtx.get_y() } });

	return json;
}

void Rectangle::showInfo(std::ostream& os) {
	os << "Name: " << name << std::endl <<
		"Type: " << "Rectangle" << std::endl <<
		"Square: " << square() << std::endl <<
		"Vertices:" << std::endl;
	for (auto& vtx : *this)
		os << "  " << vtx << std::endl;
	os << "Side_w: " << side_w << std::endl <<
		"Side h: " << side_h << std::endl;
}

File: Coursework-WinForms/Rectangle.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace Coursework_WinForms {
	public class Rectangle : Shape {
		//properties
		[JsonIgnore] public float side_w { get; private set; }
		[JsonIgnore] public float side_h { get; private set; }

		// ctors
		public Rectangle(string name, Vertex vtx, float sW, float sH) :
			base(name, vtx,
				new Vertex(vtx.x, vtx.y + sH),
				new Vertex(vtx.x + sW, vtx.y + sH),
				new Vertex(vtx.x + sW, vtx.y))
		{
			if (sH <= 0 || sW <= 0)
				throw new ArgumentException("Side lengths must be > 0");
			side_w = sW;
			side_h = sH;
		}

		public Rectangle(string name, List<Vertex> vertices) : base(name, vertices) {
			checkAndSet();
		}
		public Rectangle(string name, params Vertex[] vertices) : base(name, vertices) {
			checkAndSet();
		}

		public Rectangle(Rectangle other) : base(other.name, other.vertices) {
			checkAndSet();
			side_h = other.side_h;
			side_w = other.side_w;
		}

		// methods
		void checkAndSet() {
			if ((this[2] - this[0]).abs() != (this[3] - this[1]).abs())
				throw new ArgumentException($"{name}: The given points don't form a rectangle");
			
			side_h = (this[1] - this[0]).hypot();
			side_w = (this[2] - this[1]).hypot();
		}

		// abstract methods
		public override object Clone() {
			return new Rectangle(name, vertices);
		}

		public override double square() {
			return side_h * side_w;
		}

		public override JsonNode makeJson() {
			return JsonSerializer.SerializeToNode(this);
		}

		public override string ToString() {
			return base.ToString() +
				$"side_w: {side_w}\n" +
				$"side_h: {side_h}\n";
		}
	}
}
